pkg/restore-util/split.go

```diff
@@ -40,50 +40,74 @@ func NewRegionSplitter(client Client) *RegionSplitter {
-type OnSplitFunc func(*Range)
-
 // Split executes a region split. It will split regions by the rewrite rules,
 // then it will split regions by the end key of each range.
 // tableRules includes the prefix of a table, since some ranges may have
 // a prefix with record sequence or index sequence.
 // note: all ranges and rewrite rules must have raw key.
-func (rs *RegionSplitter) Split(
-	ctx context.Context,
-	ranges []Range,
-	rewriteRules *RewriteRules,
-	onSplit OnSplitFunc,
-) error {
+func (rs *RegionSplitter) Split(ctx context.Context, ranges []Range, rewriteRules *RewriteRules) error {
+	if len(ranges) == 0 {
+		return nil
+	}
 	startTime := time.Now()
-	rangeTree, err := newRangeTreeWithRewrite(ranges, rewriteRules)
+	// Sort the range for getting the min and max key of the ranges
+	sortedRanges, err := sortRanges(ranges, rewriteRules)
 	if err != nil {
 		return errors.Trace(err)
 	}
-	scatterRegions, err := rs.splitByRewriteRules(
-		ctx, rewriteRules.Data, onSplit)
-	if err != nil {
-		return errors.Trace(err)
+	minKey := codec.EncodeBytes([]byte{}, sortedRanges[0].StartKey)
+	maxKey := codec.EncodeBytes([]byte{}, sortedRanges[len(sortedRanges)-1].EndKey)
+	for _, rule := range rewriteRules.Table {
+		if bytes.Compare(minKey, rule.GetNewKeyPrefix()) > 0 {
+			minKey = rule.GetNewKeyPrefix()
+		}
+		if bytes.Compare(maxKey, rule.GetNewKeyPrefix()) < 0 {
+			maxKey = rule.GetNewKeyPrefix()
+		}
 	}
-	rangeTree.Ascend(func(rg *Range) bool {
-		if rg == nil {
-			return false
+	for _, rule := range rewriteRules.Data {
+		if bytes.Compare(minKey, rule.GetNewKeyPrefix()) > 0 {
+			minKey = rule.GetNewKeyPrefix()
 		}
-		if onSplit != nil {
-			onSplit(rg)
+		if bytes.Compare(maxKey, rule.GetNewKeyPrefix()) < 0 {
+			maxKey = rule.GetNewKeyPrefix()
 		}
-
-		var newRegion *RegionInfo
-		newRegion, err = rs.maybeSplitRegion(ctx, rg)
+	}
+	interval := SplitRetryInterval
+	var scatterRegions []*RegionInfo
+SplitRegions:
+	for i := 0; i < SplitRetryTimes; i++ {
+		regions, err := rs.client.ScanRegions(ctx, minKey, maxKey, 0)
 		if err != nil {
-			return false
+			return errors.Trace(err)
 		}
-		if newRegion != nil {
-			scatterRegions = append(scatterRegions, newRegion)
+		if len(regions) == 0 {
+			log.Warn("cannot scan any region")
+			return nil
 		}
-		return true
-	})
-	if err != nil {
-		return errors.Trace(err)
+		splitKeyMap := getSplitKeys(rewriteRules, sortedRanges, regions)
+		regionMap := make(map[uint64]*RegionInfo)
+		for _, region := range regions {
+			regionMap[region.Region.GetId()] = region
+		}
+		scatterRegions = make([]*RegionInfo, 0)
+		for regionID, keys := range splitKeyMap {
+			var newRegions []*RegionInfo
+			newRegions, err = rs.splitAndScatterRegions(ctx, regionMap[regionID], keys)
+			if err != nil {
+				interval = 2 * interval
+				if interval > SplitMaxRetryInterval {
+					interval = SplitMaxRetryInterval
+				}
+				time.Sleep(interval)
+				if i > 3 {
+					log.Warn("splitting regions failed, retry it")
```

Comment:
```suggestion
					log.Warn("splitting regions failed, retry it", zap.Error(err))
```

(maybe include more information)